### ISSUE TITLE
include_role: causing subsequent dependencies to run more than once.

### ISSUE THREAD

**Issue by rahul-dhumal** [ansible/ansible#24326]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

_include_role:_

##### ANSIBLE VERSION
ansible 2.2.0.0
  config file = ../playbooks/ansible.cfg
  configured module search path = Default w/o overrides

##### CONFIGURATION
No changes in ansible  configuration file

##### OS / ENVIRONMENT

Ubuntu 16.04

##### SUMMARY
When **_include_role:_** is used (specifically at the end of a .yml) file, it causes that yml file to run repeatedly more than once in all subsequent dependencies. 

##### STEPS TO REPRODUCE
[ansible-reported-bug.zip](https://github.com/ansible/ansible/files/979172/ansible-reported-bug.zip)

Attached some sample roles/playbooks which will reproduce the scenario - 
1. Unzip the include_role-bug.zip
2. run playbook main_calling.yml >>> ansible-playbook main-calling.yml

##### DEPENDENCIES IN ATTACHED SAMPLE
**-main-calling.yml** [Main playbook to run]
    - includes  **[common-test-role]**
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- **[common-test-role]** includes role - **[common-role-includes-this-another-role]** _this causes the problem_
    - includes another role as **[first-role-dependent-on-common-test-role]** which internally depends on **[common-test-role]** so common-test-role is running twice when include_role is used inside **[common-test-role]**  at the end, not anywhere else [i.e. in the beginning or in the middle]

> To Summarize : 
> Tasks directly inside **common-test-role** are executed twice when **include_role: some_role** is the last line

##### EXPECTED RESULTS
```
PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [common-test-role : debug] ************************************************
ok: [localhost] => {
    "msg": "step 1 - first inner task of common-test-role"
}

TASK [common-test-role : debug] ************************************************
ok: [localhost] => {
    "msg": "step 2 - second inner task of common-test-role"
}

TASK [just-another-role-to-reflect-internal-dependecy : debug] *****************
ok: [localhost] => {
    "msg": "step 3 - inner task of just-another-role-to-reflect-internal-dependecy [this would be displayed when common-test-role hits include_role]"
}

TASK [first-role-dependent-on-common-test-role : debug] ************************
ok: [localhost] => {
    "msg": "step 4 inside inner task of first-role-dependent-on-common-test-role"
}

PLAY RECAP *********************************************************************
localhost                  : ok=7    changed=0    unreachable=0    failed=0   

```

##### ACTUAL RESULTS

```
PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [common-test-role : debug] ************************************************
ok: [localhost] => {
    "msg": "step 1 - first inner task of common-test-role"
}

TASK [common-test-role : debug] ************************************************
ok: [localhost] => {
    "msg": "step 2 - second inner task of common-test-role"
}

TASK [just-another-role-to-reflect-internal-dependecy : debug] *****************
ok: [localhost] => {
    "msg": "step 3 - inner task of just-another-role-to-reflect-internal-dependecy [this would be displayed when common-test-role hits include_role]"
}

TASK [common-test-role : debug] ************************************************
ok: [localhost] => {
    "msg": "step 1 - first inner task of common-test-role"
}

TASK [common-test-role : debug] ************************************************
ok: [localhost] => {
    "msg": "step 2 - second inner task of common-test-role"
}

TASK [first-role-dependent-on-common-test-role : debug] ************************
ok: [localhost] => {
    "msg": "step 4 inside inner task of first-role-dependent-on-common-test-role"
}

PLAY RECAP *********************************************************************
localhost                  : ok=7    changed=0    unreachable=0    failed=0   

TASK: setup ------------------------------------------------------------- 0.40s
TASK: common-test-role : debug ------------------------------------------ 0.02s
TASK: first-role-dependent-on-common-test-role : debug ------------------ 0.02s
TASK: just-another-role-to-reflect-internal-dependecy : debug ----------- 0.02s
TASK: common-test-role : debug ------------------------------------------ 0.01s
TASK: common-test-role : debug ------------------------------------------ 0.01s
TASK: common-test-role : debug ------------------------------------------ 0.01s

Playbook finished: Fri May  5 14:20:33 2017, 7 total tasks.  0:00:00 elapsed. 


```

**Comment by jctanner** [comment]
@rahul-dhumal does this force the behavior you expect?

```
  4 - include_role:
  5     name: common-role-includes-this-another-role
  6     allow_duplicates: no 
```

needs_info

**Comment by rahul-dhumal** [comment]
I believe allow_duplicates has default value set to no so this may not work, anyways, I will try once again by explicitly setting it to 'no' & keep you posted here.

**Comment by rahul-dhumal** [comment]
Tried this but as mentioned earlier, it did not work!

```
- include_role:
	name: common-role-includes-this-another-role
	allow_duplicates: no 
```

**Comment by lhoss** [comment]
probably relates to https://github.com/ansible/ansible/issues/18748  ( and dupe: https://github.com/ansible/ansible/issues/21279 )

**Comment by sivel** [comment]
This issue was resolved in ef96329ae2279b3480261b8b737a8d41da663e5e and included in the v2.4.1 release.

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.freenode.net
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project
